tcljava/doc/VarTrace.3
'\"
'\" Copyright (c) 1998 Sun Microsystems, Inc.
'\"
'\" See the file "license.terms" for information on usage and redistribution
'\" of this file, and for a DISCLAIMER OF ALL WARRANTIES.
'\" 
'\" RCS: @(#) $Id: VarTrace.3,v 1.1 1998/10/14 21:09:09 cvsadmin Exp $
'\" 
.so man.macros
.TH VarTrace 3 1.0 Jacl "Jacl Library Procedures"
.BS
.SH INTERFACE
tcl.lang.VarTrace -- Interface for montoring the access on Tcl variables.

.SH METHODS
.sp
abstract void \fBtraceProc\fR(Interp \fIinterp\fR, String \fIname1\fR,
String \fIname2\fR, int \fIflags\fR)

.SH ARGUMENTS
.AS TclObject "table[]" .
.AP Interp interp
The interpreter accessing the variable.
.AP String name1
If the trace occurs on a scalar variable, \fIname1\fR contains the
name of the variable. If the trace occurs on an array element, or on a
whole array, \fIname1\fR contains the name of the array.
.AP String name2
If the trace occurs on an array element, \fIname2\fR contains the index of
the element. Otherwise \fIname2\fR is \fBnull\fR.
.AP int flags
Specifies when action has be applied on the variable when the trace
occurs. See below for details.
.BE

.SH DESCRIPTION

The \fBVarTrace\fR interface is used for montoring the access on Tcl
variables. A \fBVarTrace\fR instance can be registered for a Tcl
variable using \fIinterp\fR.\fBtraceVar\fR. The \fIflags\fR parameter to
\fIinterp\fR.\fBtraceVar\fR specifies which type(s) of access on the
variable should cause the \fBVarTrace\fR instance to be activated.

When the \fBVarTrace\fR instance is activated, its \fBtraceProc\fR
method is called. \fIname1\fR and \fIname2\fR give the name of the
traced variable in the normal two-part form. \fIflags\fR is an OR-ed
combination of bits providing several pieces of information. One of
the bits \fBTCL\fR.\fBTRACE_READS\fR, \fBTCL\fR.\fBTRACE_WRITES\fR, or
\fBTCL\fR.\fBTRACE_UNSETS\fR will be set in \fIflags\fR to indicate which
operation is being performed on the variable. The bit
\fBTCL\fR.\fBGLOBAL_ONLY\fR will be set whenever the variable being
accessed is a global one not accessible from the current level of procedure
call: \fBtraceProc\fR will need to pass this flag back to
variable-related procedures like \fIinterp\fR.\fBgetVar\fR if it attempts
to access the variable. The bit \fBTCL\fR.\fBTRACE_DESTROYED\fR will be set
in \fIflags\fR if the trace is about to be destroyed; this information
may be useful to \fBtraceProc\fR so that it can clean up its own
internal data structures (see the section \fBTCL\fR.\fBTRACE_DESTROYED\fR
below for more details). Lastly, the bit \fBTCL\fR.\fBINTERP_DESTROYED\fR
will be set if the entire interpreter is being destroyed. When this
bit is set, \fBtraceProc\fR must be especially careful in the things
it does (see the section \fBTCL\fR.\fBINTERP_DESTROYED\fR below).

.SH "ACCESSING VARIABLES DURING TRACES"

During read and write traces, \fBtraceProc\fR can read, write, or
unset the traced variable using \fIinterp\fR.\fBgetVar\fR,
\fIinterp\fR.\fBsetVar\fR, and other procedures.  While
\fBtraceProc\fR is executing, traces are temporarily disabled for the
variable, so that calls to \fIinterp\fR.\fBgetVar\fR and
\fIinterp\fR.\fBsetVar\fR will not cause \fBtraceProc\fR or other
trace procedures to be invoked again.  Disabling only occurs for the
variable whose trace procedure is active; accesses to other variables
will still be traced.  However, if a variable is unset during a read
or write trace then unset traces will be invoked.
 
During unset traces the variable has already been completely expunged.
It is possible for the trace procedure to read or write the variable,
but this will be a new version of the variable.  Traces are not
disabled during unset traces as they are for read and write traces,
but existing traces have been removed from the variable before any
trace procedures are invoked. If new traces are set by unset trace
pro- cedures, these traces will be invoked on accesses to the variable
by the trace procedures.
 
.SH ERROR RETURNS

If \fBtraceProc\fR throws a TclException, it signifies that an error
occurred, in which case no further traces are invoked for the access
and the traced access aborts with the message given by the
TclException. Trace procedures can use this facility to make variables
read-only, for example (but note that the value of the variable will
already have been modified before the trace procedure is called, so
the trace procedure will have to restore the correct value).
 
The TclException is used only during read and write tracing. During
unset traces, the TclException is ignored and all relevant trace
procedures will always be invoked.

.SH TCL.TRACE_DESTROYED FLAG

In an unset callback to \fBtraceProc\fR, the
\fBTCL\fR.\fBTRACE_DESTROYED\fR bit is set in \fIflags\fR if the trace
is being removed as part of the deletion. Traces on a variable are
always removed whenever the variable is deleted; the only time
\fBTCL\fR.\fBTRACE_DESTROYED\fR isn't set is for a whole-array trace
invoked when only a single element of an array is unset.
 
.SH TCL.INTERP_DESTROYED FLAG

When an interpreter is destroyed, unset traces are called for all of
its variables. The \fBTCL\fR.\fBINTERP_DESTROYED\fR bit will be set in the
\fIflags\fR argument passed to the trace procedures. Trace procedures
must be extremely careful in what they do if the
\fBTCL\fR.\fBINTERP_DESTROYED\fR bit is set. It is not safe for the
procedures to invoke any Tcl procedures on the interpreter, since its
state is partially deleted.  All that trace procedures should do under
these circumstances is to clean up and free their own internal data
structures.

.SH "EQUIVALENT C FUNCTIONS"
Tcl_TraceProc

.SH "SEE ALSO"
Interp, setVar, traceVar

.SH KEYWORDS
trace, variable


tcljava/ChangeLog
2006-06-29  Mo DeJong  <mdejong@users.sourceforge.net>

        * README.jacl:
        * README.tclblend: Mention that Jacl and Tcl Blend
        are production ready and include URLs for build
        notes on the wiki.

2006-06-29  Mo DeJong  <mdejong@users.sourceforge.net>

        Update version numbers to 1.3.3.

        * configure:
        * configure.in:
        * src/pkgIndex.tcl:
        * src/jacl/tcl/lang/Interp.java:
        * src/tcljava/tcl/lang/BlendExtension.java:
        * unix/tools/vutil/current:
        * win/makefile.vc:

2006-06-29  Mo DeJong  <mdejong@users.sourceforge.net>

        * configure: Regen.
        * configure.in: Set default --prefix to /usr/local.

2006-04-05  Mo DeJong  <mdejong@users.sourceforge.net>

        * src/tclblend/tcl/lang/Interp.java (evalFile):
        Add TclBlend implementation of Interp.evalFile().

2006-04-05  Mo DeJong  <mdejong@users.sourceforge.net>

        * src/jacl/tcl/lang/ExecCmd.java (cmdProc):
        Reimplement JDK 1.3 and newer exec so that two
        threads are created. One reads stdout of the
        subprocess while the other reads stderr. This
        is needed to avoid deadlocking in waitFor()
        when a process writes a lot of stdout and stderr.
        * src/tests/tcljava/tests/exec/ExecLotsOutput.java:
        * tests/inprogress/ExecCmd.test: Add test case
        for the waitFor() exec deadlocking bug.

2005-08-08  Mo DeJong  <mdejong@users.sourceforge.net>

	Add tcljava-132-branch branch and
 	and tcljava-132-branchpoint tags.

2005-08-08  Mo DeJong  <mdejong@users.sourceforge.net>

	Add tcljava-132-2005-08-08 tag to mark tree
	state for the 1.3.2 release.

2005-07-28  Mo DeJong  <mdejong@users.sourceforge.net>

	* announce: Update.
	* docs/website/index.html: Update.

2005-07-27  Mo DeJong  <mdejong@users.sourceforge.net>

	* README.jacl: Update msys description and
	mention jaclsh.bat script.
	* README.tclblend: Mention Tcl 8.4.11 and
	Thread 2.6.2 reqs. Add msys url and
	mention jtclsh.bat script.
	* docs/website/faq.html: Update FAQ.
	* docs/website/index.html: Update homepage
	info with new 1.3.2 release.
	* docs/website/mail.html: Add mail page that
	explains how the mailing lists work.

2005-07-23  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/Channel.java (initOutput):
	When initializing a channel of type "file",
	set a flag to indicate that the channel
	should be sync()'ed in addition to flushing.
	* src/jacl/tcl/lang/FlushCmd.java: Print
	IO exception stack trace, this should never
	happen but if it did sa stack trace would
	be helpful.
	* src/jacl/tcl/lang/TclOutputStream.java (setSync,
	flushChannel):
	Add flag setting method so that an output
	stream can be marked so that sync() will
	be invoked when flushing.
	* tests/tcl/SocketCmd.test: Add simple socket
	test case that writes to and reads from a socket.

2005-07-21  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/BinaryCmd.java (FormatNumber,
	ScanNumber):
	Handle edge cases when casting from a double
	to a float. Rewrite scanning code so that
	64 bit long value is used for shifting operation.
	* src/jacl/tcl/lang/InfoCmd.java:
	* src/jacl/tcl/lang/Shell.java:
	* src/jacl/tcl/lang/SubstCmd.java:
	Invoke static method via class name.
	* src/jacl/tcl/lang/Util.java (strtod): Add
	support for "NaN", "Inf", and "-Inf" strings
 	as valid double values.
	* tests/tcl/binary.test: Enable test cases
	that were not being run because they had
	the nonPortable constraint. Java supports
	only big endian primitive types so the
	binary format and scan operations can
	just assume big endian layout. Test that
	assume little endian are not run via
	addition of the littleEndian constraint.
	Add test cases for double -> float edge cases.

2005-07-20  Mo DeJong  <mdejong@users.sourceforge.net>

	* Makefile.in: Copy new files into dist.
	* tests/tclblend/loading.test: Rework test
	so that it is only run under Win32 and
	returns the proper result.
	* win/jaclsh.bat: Removed old bat script.

2005-07-19  Mo DeJong  <mdejong@users.sourceforge.net>

	* Makefile.in: Remove old green threads code
	for "custom" target. Tcl Blend only works
	with a native threads JVM anyway. Add new
	custom2 shell which loads a JVM at runtime
	and then loads Tcl Blend like a JVM would.
	This shell make it easier to debug weird
	loading issues in Tcl Blend.
	* src/tests/tclblend/tests/JavaEval.java:
	Make the program exit(0) when an expected
	result matches the program result. Use
	positive exit status.
	* tests/tclblend/loading.test: Use the
	expected compare test in the JavaEval
	class.
	* unix/custom2.c: Add new custom shell impl
	that loads a JVM and then loads Tcl Blend into
	the JVM.

2005-07-18  Mo DeJong  <mdejong@users.sourceforge.net>

	Fix crashes when loading Tcl Blend into an
	existing JVM by invoking Tcl_FindExecutable("java")
	to init Tcl subsystems.

	* src/native/javaCmd.c: Update comment.
	* src/native/javaInterp.c (
	Java_tcl_lang_Interp_create,
	Java_tcl_lang_Interp_init,
	Java_tcl_lang_Interp_initName): Add
	TCLBLEND_DEBUG output statements.
	Add initName() method that will init the
	Tcl executable name when Tcl Blend is loaded
	from Java.
	* src/tclblend/tcl/lang/Interp.java (shlib_load,
	initName):
	Add initName() method mentioned above and
	invoke it after loading the Tcl Blend shared lib.
	* tests/tclblend/loading.test: Update tests
	to account for executable name check. Add
	test for [info nameofexecutable] in a Tcl Blend
	interp loaded into Java.

2005-07-17  Mo DeJong  <mdejong@users.sourceforge.net>

	Add startup .bat files for Jacl and Tcl Blend.
	The bat files are auto generated by configure
	with the correct paths.

	* Makefile.in: Install and test jaclsh.bat
	and jtclsh.bat. Make sure the PATH variable
	is always set with the new directories
	prepended to the PATH. Rework tricky gdb
	invocation under Win32 since the mingw
	version of gdb does not know about mingw
	style unix paths.

tcljava/src/tcljava/tcl/lang/VarTrace.java
/*
 * VarTrace.java --
 *
 *	Interface for creating variable traces.
 *
 * Copyright (c) 1997 Sun Microsystems, Inc.
 *
 * See the file "license.terms" for information on usage and
 * redistribution of this file, and for a DISCLAIMER OF ALL
 * WARRANTIES.
 * 
 * RCS: @(#) $Id: VarTrace.java,v 1.1 1998/10/14 21:09:14 cvsadmin Exp $
 *
 */

package tcl.lang;

/*
 * This interface is used to make variable traces. To make a variable
 * trace, write a class that implements the VarTrace and call
 * Interp.traceVar with an instance of that class.
 * 
 */

public interface VarTrace {


/*
 *----------------------------------------------------------------------
 *
 * traceProc --
 *
 *	This function gets called when a variable is accessed.
 *
 * Results:
 *	None.
 *
 * Side effects:
 *	The traceProc can cause arbitrary side effects. If a
 *	TclException is thrown, error message is stored in the result
 *	of the interpreter.
 *
 *----------------------------------------------------------------------
 */

abstract public void
traceProc(
    Interp interp,		// Current interpreter.
    String part1,		// First part of the variable name.
    String part2,		// Second part of the var name. May be null.
    int flags)			// TCL.TRACE_READS, TCL.TRACE_WRITES or
				// TCL.TRACE_UNSETS (exactly one of these
				// bits will be set.)
throws
     TclException;		// The traceProc may throw a TclException
				// to indicate an error during the trace.

} // end VarTrace

